### ISSUE TITLE
Mac build problem: Mismatch between OS and SDK versions

### ISSUE THREAD

**Issue by hendrikweisser** [OpenMS/OpenMS#1829]
I'm mainly documenting this in case someone else has the same problem. The OS on my MacBook was updated to 10.10.5 and I installed Xcode 7 to get the developer tools ("make" etc.). Then I got weird linker errors when I tried to rebuild my OpenMS:
`Undefined symbols for architecture x86_64: [...]`

So I ran "make clean", removed the "CMakeCache.txt" and attempted a fresh build. However, the command from the documentation didn't work:
```
$ cmake -DCMAKE_PREFIX_PATH="/Users/hw5/Software/OpenMS-contrib/build" ..
CMake Warning at /Users/hw5/Software/Homebrew/Cellar/cmake/3.0.2/share/cmake/Modules/Platform/Darwin.cmake:179 (message):
  Ignoring CMAKE_OSX_SYSROOT value:

   /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk

  because the directory does not exist.
[...]
CMake Error at /Users/hw5/Software/Homebrew/Cellar/cmake/3.0.2/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:136 (message):
  Could NOT find OpenGL (missing: OPENGL_INCLUDE_DIR)
```

Then I updated CMake via Homebrew and the second part of the error changed:
```
-- Check for working C compiler: /usr/bin/cc
CMake Error at /Users/hw5/Software/Homebrew/Cellar/cmake/3.4.3/share/cmake/Modules/Platform/Darwin.cmake:76 (message):
  CMAKE_OSX_DEPLOYMENT_TARGET is '10.10' but CMAKE_OSX_SYSROOT:

   ""

  is not set to a MacOSX SDK with a recognized version.  Either set
  CMAKE_OSX_SYSROOT to a valid SDK or set CMAKE_OSX_DEPLOYMENT_TARGET to
  empty.
```

After some googling, I got it to work using this:
`cmake -DCMAKE_PREFIX_PATH="/Users/hw5/Software/OpenMS-contrib/build" -DCMAKE_OSX_DEPLOYMENT_TARGET=10.10 -DCMAKE_OSX_SYSROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk ..`

Note the different version numbers for the OS and the SDK.

**Comment by hendrikweisser** [comment]
Now I actually need some help:
The OpenMS build failed again with a long "Undefined symbols for architecture x86_64" error seemingly related to boost. So I decided to rebuild the contrib package (completely, i.e. "BUILD_TYPE=ALL"). Again I had to set the "CMAKE_OSX_DEPLOYMENT_TARGET" and "CMAKE_OSX_SYSROOT" variables. However, it seems that these variables are not passed on when the actual libraries are built - I now get this error:
```
CMake Error at libraries.cmake/seqan.cmake:39 (message):
  Generating seqan build system ..  failed
```
The "contrib_build.log" file shows the same errors related to the OS and SDK versions that I got before, but now for the SeqAn build.
How do I get CMake to pass the variables on to the library build calls?

**Comment by hendrikweisser** [comment]
Apparently the "CMAKE_OSX_SYSROOT" variable can be initialized based on an environment variable, "SDKROOT" (see https://cmake.org/cmake/help/v3.0/variable/CMAKE_OSX_SYSROOT.html). However, setting "SDKROOT" didn't change anything for me.

I finally got SeqAn to build by adding `-D CMAKE_OSX_SYSROOT=/` to the CMake options in "libraries.cmake/seqan.cmake" inside the contrib directory. All the other contrib libraries built successfully without changes.

**Comment by hendrikweisser** [comment]
The contrib package is built, but OpenMS still fails with Boost-related linker errors. That is with "BOOST_USE_STATIC=ON" in the CMake options. If I switch that to "OFF", I instead get linker errors ("Undefined symbols..." as before) for WildMagic. Interesting.

**Comment by jpfeuffer** [comment]
Weird. Restarted our builds on Yosemite, ElCapitan and Mavericks. Whole contrib builds fine. OpenMS is to come.

**Comment by hendrikweisser** [comment]
I forgot to mention, I downgraded to Xcode 6 to get rid of the OS/SDK mismatch. I now have the 10.10 SDK, but CMake still wouldn't work without setting "CMAKE_OSX_SYSROOT" explicitly!

@jpfeuffer:
What CMake version do you use?

**Comment by jpfeuffer** [comment]
Mavericks: 3.3.1
El Capitan: 3.2.3
Yosemite: 3.2.3

**Comment by hendrikweisser** [comment]
@jpfeuffer:
I'm on Yosemite with CMake 3.4.3. Could you send me the CMake settings (what "ccmake" prints) for your Yosemite test system?

**Comment by hendrikweisser** [comment]
Does anyone have an idea what I could try to solve or diagnose this problem? I would like to get OpenMS running again on my laptop before the retreat.

**Comment by jpfeuffer** [comment]
Hey sorry, I was very busy the last week with two conferences and today I just fixed our "beta" OSX build servers again. I will report on the configuration this evening or tomorrow morning and hope that it helps you.

**Comment by jpfeuffer** [comment]
Ok. Contrib first:
We run (from contrib/build ) 
`cmake .. -DBUILD_TYPE=ALL -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_C_COMPILER=clang`

when I grep the build_log for OSX:
```
-- Found OpenGL: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/OpenGL.framework
-- Found GLUT: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/GLUT.framework
  Policy CMP0042 is not set: MACOSX_RPATH is enabled by default.  Run "cmake
  MACOSX_RPATH is not specified for the following targets:
     CMAKE_OSX_ARCHITECTURES
     CMAKE_OSX_DEPLOYMENT_TARGET
     CMAKE_OSX_SYSROOT
```
grep OSX in CMakeCache:
```
//Build architectures for OSX
CMAKE_OSX_ARCHITECTURES:STRING=
CMAKE_OSX_DEPLOYMENT_TARGET:STRING=
CMAKE_OSX_SYSROOT:PATH=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk
//building multiple architectures on MacOSX
OPENMS_CONTRIB_MACOSX_MIXED_MODE:INTERNAL=0
```

**Comment by jpfeuffer** [comment]
Btw: I stumbled across these lines while digging through the cmake scripts.
https://github.com/OpenMS/OpenMS/blob/develop/CMakeLists.txt#L55

Might be related to your problem.

**Comment by hendrikweisser** [comment]
@jpfeuffer: Thanks.

> CMAKE_OSX_SYSROOT:PATH=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk

Is that on the 10.10 system?

> Could you send me the CMake settings (what "ccmake" prints) for your Yosemite test system?

Again, can you please post the whole configuration?

**Comment by hendrikweisser** [comment]
> Btw: I stumbled across these lines while digging through the cmake scripts.

I commented out those lines, and now I get the same warning about "Policy CMP0042" that you showed above, but no change for the actual build.

**Comment by jpfeuffer** [comment]
Here the full CMakeCache.txt (a slightly more verbose ccmake output) on 10.11.2:
https://gist.github.com/jpfeuffer/03620da581d41c469741

The following contains the same for the 10.10.5 system:
https://gist.github.com/jpfeuffer/8b6a00c111a240526bf4

Seems like the use the same Sysroot path, because they both have:
Xcode 7.2, Build version 7C68
installed.

**Comment by jpfeuffer** [comment]
I found another thing to play with:
On our very old OSX machine in Berlin (OSX 10.8.5, Xcode 5.1.1) we use the following script to run the builds:
https://github.com/OpenMS/build-scripts/blob/master/nightlies/osx/run_tests.cmake

Interesting for you might be this part.

```
set(INITIAL_CACHE
	"
...
CMAKE_OSX_SYSROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.8.sdk/
CMAKE_OSX_DEPLOYMENT_TARGET=10.6
")
```
Maybe your settings did not propagate because they were not really in the CMake cache?

**Comment by hendrikweisser** [comment]
The underlying problem seems to have been that the (dynamic) linker could not find all the necessary libraries. Running "otool -L" on a TOPP tool showed that for contrib libraries, only the name was known, not the path to library (like for other libraries). Why that happened and how it could be fixed is unclear.

A solution was to install libraries (e.g. Boost) through Homebrew, and for those that aren't available through Homebrew (e.g. WildMagic), set the environment variable "DYLD_LIBRARY_PATH" to the contrib's "lib" directory. This allows the OS to find the dynamic libraries, and avoids the following error:
```
dyld: Library not loaded: libWm5Core.dylib
  Referenced from: [TOPP tool]
  Reason: image not found
Trace/BPT trap: 5
```

A remaining issue is that TOPPView and TOPPAS cannot be started through the app bundles (because "DYLD_LIBRARY_PATH" is not set in this case), but they can be used from the command line.

**Comment by timosachsenberg** [comment]
what is the state here? does this have any implications on the release in your view?

**Comment by hendrikweisser** [comment]
> what is the state here?

As described above, I got it (mostly) working on my Mac by setting "DYLD_LIBRARY_PATH". I haven't found the root cause of the problem, but don't have time to investigate it further.

> does this have any implications on the release in your view?

Since this will be a source release, we really should make sure that people can build OpenMS by following the instructions we provide. So we should have as many people test that as possible. However, so far nobody else has come upon the same problem as I did, so perhaps there's just something weird about my Mac configuration.
